components/ObjectDisplay.js

```diff
@@ -0,0 +1,201 @@
+      }}
+    >
+      <div
+        onClick={onClick}
+        style={{
+          cursor: 'pointer',
+          position: 'relative'
+        }}
+      >
+        <Arrow style={{ left: '-16px' }} />
+        <span>
+          {!!rootKey && (
+            <>
+              <span style={{ color: COLORS.string }}>"{rootKey}"</span>
+              {': '}
+            </>
+          )}
+          {leading}
+        </span>
+      </div>
+      <div style={{ marginLeft: 16 }}>{children}</div>
+      <span>{`${trailing}${hasTrailingComma ? ',' : ''}`}</span>
+    </pre>
+  );
+};
+
+const isObject = (value) =>
+  !Array.isArray(value) && typeof value === 'object' && value !== null;
+
+export const ObjectDisplay = ({ data, rootKey, isRoot, minCollapseSize }) => {
+  const childCount = Object.keys(data).length;
+  const isEmpty = childCount === 0;
+
+  const [isCollapsed, setCollapsed] = React.useState(
+    () => !isRoot && childCount > minCollapseSize
+  );
+
+  if ((isObject(data) || Array.isArray(data)) && (isCollapsed || isEmpty)) {
+    const [leading, trailing] = getSurroundingChars(Array.isArray(data));
+
+    return (
+      <KeyValueRow
+        onClick={isEmpty ? null : () => setCollapsed(false)}
+        canCopy={isRoot}
+        data={data}
+      >
+        {!!rootKey && (
+          <>
+            <span style={{ color: COLORS.string }}>"{rootKey}"</span>
+            <pre>{': '}</pre>
+          </>
+        )}
+        <pre style={{ color: COLORS.default }}>
+          {isEmpty ? (
+            `${leading}${trailing}`
+          ) : (
+            <>
+              {leading}
+              <span
+                style={{ color: 'rgb(163, 172, 185)', lineHeight: 0 }}
+              >{` ... `}</span>
+              <span
+                style={{ lineHeight: 0, color: 'rgb(163, 172, 185)' }}
+              >{`${childCount} item${childCount === 1 ? '' : 's'} `}</span>
+              {trailing}
+            </>
+          )}
+        </pre>
+        {!isRoot && <pre>{','}</pre>}
+      </KeyValueRow>
+    );
+  }
+
+  const lines = Object.entries(data).map(([key, value]) => {
+    const valueColor = COLORS[typeof value] ?? COLORS.default;
+
+    if (isObject(value) || Array.isArray(value)) {
+      return (
+        <ObjectDisplay
+          data={value}
+          rootKey={key}
+          minCollapseSize={minCollapseSize}
+        />
+      );
+    }
+
+    return (
+      <KeyValueRow key={key} hasTrailingComma={!isRoot}>
+        {isObject(data) && (
+          <>
+            <pre style={{ color: COLORS.string }}>"{key}"</pre>
+            <pre>{': '}</pre>
+          </>
+        )}
+        {!isObject(value) &&
+        !Array.isArray(value) && ( // Can probably remove this check...
+            <pre style={{ color: valueColor }}>
+              {typeof value === 'string' ? `"${value}"` : String(value)}
+            </pre>
+          )}
+        <pre>{','}</pre>
+      </KeyValueRow>
+    );
+  });
+
+  return (
```

Comment:
Thanks @ianjabour-stripe for the inspiration here!

But maybe a better solution here is use code mirror folding? - https://codemirror.net/demo/folding.html